### PR TITLE
LibWeb: Implement support for DOM::Document in XHR::send()

### PR DESCRIPTION
**Pull request by kennethmyhra**
Made sure to verify that osnews.com loads so Andreas still can read his OS news in Ladybird :^)

### CODE REVIEW

**Review by linusg — COMMENTED** [review]
Nested union support was already added in https://github.com/SerenityOS/serenity/pull/15732, no? If you think reverting that is the right solution, commit 1 + 3 should probably be combined, and needs a little more explanation why it didn't work out as intended. At a glance this still looks a bit like a hack...

**Comment by kennethmyhra** [comment]
The thing with how it was implemented in #15732 is that it flattens the nested union type so that i no longer represents the actual nested union type in C++, if that makes sense.

I agree that it is still a bit hackish, but currently I haven't found a less hacky way to implement it. It does however now resemble the nested union types in C++ and IDL the same way.

Commit 1 and 3 compiles and works independently of each other, and I kinda did that on purpose to make it more clear how the changes affect the code base. I can however squash those into one commit.

**Comment by kennethmyhra** [comment]
And of course the title here is a bit misleading, should have said `Change behavior of nested union types support` or something along those lines

**Comment by kennethmyhra** [comment]
Let me fix up the commits and then do some testing so I don't break anything

File: Userland/Libraries/LibWeb/XHR/XMLHttpRequest.cpp [hunk 1]
```diff
@@ -436,7 +436,16 @@ WebIDL::ExceptionOr<void> XMLHttpRequest::send(Optional<Fetch::XMLHttpRequestBod
...
 
-    auto body_with_type = body.has_value() ? TRY(Fetch::extract_body(realm, body.value())) : Optional<Fetch::Infrastructure::BodyWithType> {};
+    Optional<Fetch::Infrastructure::BodyWithType> body_with_type {};
+    Optional<String> serialized_document {};
+    if (body.has_value()) {
+        if (body->has<JS::Handle<DOM::Document>>()) {
+            serialized_document = TRY(body->get<JS::Handle<DOM::Document>>().cell()->serialize_fragment(DOMParsing::RequireWellFormed::No));
+        } else {
+            auto body_init_or_readable_bytes = body->downcast<JS::Handle<Streams::ReadableStream>, JS::Handle<FileAPI::Blob>, JS::Handle<JS::Object>, JS::Handle<URL::URLSearchParams>, String, ReadonlyBytes>();
+            body_with_type = TRY(Fetch::extract_body(realm, body_init_or_readable_bytes));
```
Comment (linusg) [inline Userland/Libraries/LibWeb/XHR/XMLHttpRequest.cpp:446]: CxByte kindly added this in https://github.com/SerenityOS/serenity/pull/16002:

```suggestion
            body_with_type = TRY(Fetch::extract_body(realm, body->downcast<BodyInitOrReadableBytes>()));
```

File: Userland/Libraries/LibWeb/XHR/XMLHttpRequest.idl [hunk 2]
```diff
@@ -33,7 +36,7 @@ interface XMLHttpRequest : XMLHttpRequestEventTarget {
     undefined open(DOMString method, DOMString url);
     undefined open(ByteString method, USVString url, boolean async, optional USVString? username = {}, optional USVString? password = {});
     undefined setRequestHeader(DOMString name, DOMString value);
-    undefined send(optional XMLHttpRequestBodyInit? body = null);
+    undefined send(optional DocumentOrXMLHttpRequestBodyInit? body = null);
```
Comment (linusg) [inline Userland/Libraries/LibWeb/XHR/XMLHttpRequest.idl:39]: We can write this according to the spec once https://github.com/SerenityOS/serenity/pull/16009 is merged

File: Userland/Libraries/LibWeb/XHR/XMLHttpRequest.cpp (hunk 1 above)
Comment (kennethmyhra) [reply Userland/Libraries/LibWeb/XHR/XMLHttpRequest.cpp:446]: So I can actually downcast it to a completely different Variant, CxByte really is a wizard

**Review by linusg — APPROVED** [review]
LGTM!

**Pull request merged**